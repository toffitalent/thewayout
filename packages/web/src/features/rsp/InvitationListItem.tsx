import {
  Avatar,
  Button,
  Flex,
  Menu,
  MenuGroup,
  MenuItem,
  MenuList,
  MenuTrigger,
  Modal,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalOverlay,
  ModalTitle,
  Text,
  Toast,
} from '@disruptive-labs/ui';
import DotsVerticalIcon from '@disruptive-labs/ui/dist/icons/DotsVertical';
import { useState } from 'react';
import { RspInvitation } from '@two/shared';
import { useAppDispatch, useAppSelector, useErrorIndicator } from '@app/hooks';
import { formatPhoneNumber } from '@app/utils';
import { removeInvitation } from './actions';
import { selectRsp } from './reducer';

export const InvitationListItem = (invitation: RspInvitation) => {
  const rsp = useAppSelector(selectRsp);
  const showError = useErrorIndicator();
  const dispatch = useAppDispatch();
  const [isModalOpen, setModalOpen] = useState(false);
  const { firstName, lastName, phone, email, id } = invitation;

  const handleRemove = async () => {
    if (rsp) {
      dispatch(removeInvitation({ rspId: rsp.id, invitationId: id }))
        .unwrap()
        .then(() => {
          setModalOpen(false);
          Toast.success('Invitation removed!');
        })
        .catch((e) => showError(e));
    }
  };

  return (
    <Flex direction="row" borderBottom alignItems="center">
      <Flex direction="row" width="1/3" alignItems="center" pv={5}>
        <Avatar size="xs" name={`${firstName} ${lastName}`} mr={3} />
        <Text fontWeight="700">{`${firstName} ${lastName}`}</Text>
      </Flex>
      <Text width="1/3" whiteSpace="nowrap" overflow="hidden" style={{ textOverflow: 'ellipsis' }}>
        {email}
      </Text>
      <Text width="1/3">{formatPhoneNumber(phone)}</Text>
      <Flex width={12}>
        <Menu>
          <MenuTrigger>
            <DotsVerticalIcon height={6} width={6} cursor="pointer" data-testid="menu" />
          </MenuTrigger>
          <MenuList>
            <MenuGroup>
              <MenuItem minWidth={40} onClick={() => setModalOpen(true)}>
                Delete
              </MenuItem>
            </MenuGroup>
          </MenuList>
        </Menu>
      </Flex>
      <Modal isOpen={isModalOpen} onClose={() => setModalOpen(false)}>
        <ModalOverlay />
        <ModalContent rounded="xl" p={8}>
          <ModalCloseButton color="primary" />
          <ModalTitle mb={8} f="s1">
            Are you sure you want to delete this invitation?
          </ModalTitle>
          <ModalFooter pt={0}>
            <Button colorScheme="primary" onClick={handleRemove}>
              Delete
            </Button>
          </ModalFooter>
        </ModalContent>
      </Modal>
    </Flex>
  );
};
