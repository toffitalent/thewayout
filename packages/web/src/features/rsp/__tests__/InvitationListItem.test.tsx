import { fixtures, render, waitFor } from '@test';
import { InvitationListItem } from '../InvitationListItem';

const { invitation } = fixtures;

describe('InvitationListItem', () => {
  test('renders correctly', async () => {
    const { container, findByText } = render(<InvitationListItem {...invitation} />);
    await waitFor(async () =>
      expect(
        await findByText(`${invitation.firstName} ${invitation.lastName}`),
      ).toBeInTheDocument(),
    );
    expect(container).toMatchSnapshot();
  });
});
