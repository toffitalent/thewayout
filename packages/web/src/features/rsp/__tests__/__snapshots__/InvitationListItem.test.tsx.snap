// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvitationListItem renders correctly 1`] = `
<div>
  <div
    class="borderBottom alignItems-center flexDirection-row"
  >
    <div
      class="alignItems-center flexDirection-row width-1/3 paddingTop-5 paddingBottom-5"
    >
      <div
        class="avatar s-xs marginRight-3"
      >
        <div
          aria-label="First Last"
          class="initials"
        >
          FL
        </div>
      </div>
      <p
        class="text fontWeight-700"
      >
        First Last
      </p>
    </div>
    <p
      class="text overflow-hidden width-1/3 whiteSpace-nowrap"
      style="text-overflow: ellipsis;"
    >
      first.last@email.com
    </p>
    <p
      class="text width-1/3"
    >
      (334) 222-3333
    </p>
    <div
      class="width-12"
    >
      <svg
        aria-controls="menu-:r0:"
        aria-expanded="false"
        aria-haspopup="menu"
        aria-hidden="true"
        class="icon width-6 height-6"
        cursor="pointer"
        data-testid="menu"
        focusable="false"
        id="menu-:r0:-trigger"
        role="button"
        stroke-width="2"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0z"
          stroke="none"
        />
        <path
          d="M11 12a1 1 0 102 0 1 1 0 10-2 0M11 19a1 1 0 102 0 1 1 0 10-2 0M11 5a1 1 0 102 0 1 1 0 10-2 0"
        />
      </svg>
      <div
        class="menu"
        data-popper-escaped="true"
        data-popper-placement="bottom-start"
        data-popper-reference-hidden="true"
        style="visibility: hidden; position: absolute; top: 0px; left: 0px; min-width: max-content; will-change: transform, visibility; transform: translate(0px, 0px);"
      >
        <div
          aria-orientation="vertical"
          class="list"
          id="menu-:r0:"
          role="menu"
          tabindex="-1"
        >
          <div
            class="group"
            role="group"
          >
            <button
              class="button md text c-primary item minWidth-40"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
